src/ferui-components/tree-view/tree-view-component.ts

```diff
@@ -0,0 +1,425 @@
+    this.cd.markForCheck();
+  }
+
+  /**
+   * Emits the Node Event for outside Tree View Component usage as well as ensures tree nodes properties are updated
+   * @param {TreeViewEvent<any>} event
+   */
+  private nodeEvent(event: TreeNodeEvent<T>): void {
+    this.emitTreeViewEvent(event);
+    switch (event.getType()) {
+      case TreeViewEventType.NODE_EXPANDED:
+        this.handleExpandNode(event.getNode(), this.vs.viewPortInfo.startIndex, this.vs.viewPortInfo.endIndex);
+        break;
+      case TreeViewEventType.NODE_COLLAPSED:
+        this.handleCollapseNode(event.getNode());
+        break;
+      case TreeViewEventType.NODE_CLICKED:
+        this.selectOneNode(event.getNode());
+        break;
+      default:
+        break;
+    }
+  }
+
+  /**
+   * Selects a TreeNode and deselects all other on the entire Tree View
+   * @param {TreeNode<any>} node
+   */
+  private selectOneNode(node: TreeNode<T>): void {
+    this.scrollViewArray.forEach(scrollItem => (scrollItem.selected = scrollItem === node));
+  }
+
+  /**
+   * On any change of the scroll view array we rebuild to ensure view is updated correctly
+   */
+  private rebuildVirtualScrollerArray(): void {
+    const aggregator = [];
+    if (this.rootNode) {
+      this.addNodeAndChildrenToVirtualScrollerArray(this.rootNode, aggregator);
+    } else {
+      for (const child of this.nonRootArray) {
+        this.addNodeAndChildrenToVirtualScrollerArray(child, aggregator);
+      }
+    }
+    this.scrollViewArray = aggregator;
+    this.cd.markForCheck();
+  }
+
+  /**
+   * Returns whether the child node branch is complete and all its children loaded
+   * @param {TreeNode<T>} node
+   * @param {Array<TreeNode<T>>} aggregator
+   * @returns {boolean}
+   */
+  private addNodeAndChildrenToVirtualScrollerArray(node: TreeNode<T>, aggregator: Array<TreeNode<T>>): boolean {
+    aggregator.push(node);
+    for (const child of node.children) {
+      if (!this.addNodeAndChildrenToVirtualScrollerArray(child, aggregator)) {
+        return false;
+      }
+    }
+    return node.expanded === false || node.allChildrenLoaded;
+  }
+
+  /**
+   * Gets the first child node with more children to load
+   * @param {TreeNode<T>} node
+   * @returns {TreeNode<T>}
+   */
+  private getFirstNodeWithMoreChildrenToLoad(node: TreeNode<T>): TreeNode<T> | null {
+    if (node.expanded === true && !node.allChildrenLoaded) {
+      return node;
+    }
+    for (const child of node.children) {
+      const childResult = this.getFirstNodeWithMoreChildrenToLoad(child);
+      if (childResult !== null) {
+        return childResult;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Handles the expansion of a Tree Node component
+   * @param {TreeNode<T>} node
+   * @param {number} firstIdxInView
+   * @param {number} lastIdxInView
+   * @returns {Promise<void>}
+   */
+  private async handleExpandNode(node: TreeNode<T>, firstIdxInView: number, lastIdxInView: number): Promise<void> {
```

Comment:
The thing you need to do to optimise this function is to make it a [cancelable promise](https://stackoverflow.com/questions/46461801/possible-to-add-a-cancel-method-to-promise-in-typescript?answertab=active#answer-46465935) and cancel the promise, if it already exist, each time you call this function and then re-run again for the new nodes to be loaded like you did.